File: 2016.sp1/linux/devkit/plug-ins/scripted/multiPlugInfoCmd.py
#-
# ==========================================================================
# Copyright 2010 Autodesk, Inc. All rights reserved.
#
# Use of this software is subject to the terms of the Autodesk
# license agreement provided at the time of installation or download,
# or which otherwise accompanies this software in either electronic
# or hard copy form.
# ==========================================================================
#+

#
# This plugin prints out the child plug information for a multiPlug.
# If the -index flag is used, the logical index values used by the plug
# will be returned.  Otherwise, the plug values will be returned.
#
# import maya.cmds as cmds
# cmds.loadPlugin("multiPlugInfoCmd.py")
# cmds.multiPlugInfo("myObj.myMultiAttr", index=True)
#

import maya.OpenMaya as OpenMaya
import maya.OpenMayaMPx as OpenMayaMPx
import sys
kPluginCmdName  = "multiPlugInfo"
kIndexFlag		= "-i"
kIndexFlagLong	= "-index"

# Wrapper to handle exception when MArrayDataHandle hits the end of the array.
def advance(arrayHdl):
	try:
		arrayHdl.next()
	except:
		return False

	return True


# command
class multiPlugInfo(OpenMayaMPx.MPxCommand):
	def __init__(self):
		OpenMayaMPx.MPxCommand.__init__(self)
		# setup private data members
		self.__isIndex = False

	def doIt(self, args):
		"""
		This method is called from script when this command is called.
		It should set up any class data necessary for redo/undo,
		parse any given arguments, and then call redoIt.
		"""
		argData = OpenMaya.MArgDatabase(self.syntax(), args)

		if argData.isFlagSet(kIndexFlag):
			self.__isIndex = True

		# Get the plug specified on the command line.
		slist = OpenMaya.MSelectionList()
		argData.getObjects(slist)
		if slist.length() == 0:
			print "Must specify an array plug in the form <nodeName>.<multiPlugName>."
			return

		plug = OpenMaya.MPlug()
		slist.getPlug(0, plug)
		if plug.isNull():
			print "Must specify an array plug in the form <nodeName>.<multiPlugName>."
			return

		# Construct a data handle containing the data stored in the plug.
		dh = plug.asMDataHandle()
		adh = None
		try:
			adh = OpenMaya.MArrayDataHandle(dh)
		except:
			print "Could not create the array data handle."
			plug.destructHandle(dh)
			return

		# Iterate over the values in the multiPlug.  If the index flag has been used, just return
		# the logical indices of the child plugs.  Otherwise, return the plug values.
		for i in range(adh.elementCount()):
			try:
				indx = adh.elementIndex()
			except:
				advance(adh)
				continue

			if self.__isIndex:
				self.appendToResult(indx)
			else:
				h = adh.outputValue()
				if h.isNumeric():
					if h.numericType() == OpenMaya.MFnNumericData.kBoolean:
						self.appendToResult(h.asBool())
					elif h.numericType() == OpenMaya.MFnNumericData.kShort:
						self.appendToResult(h.asShort())
					elif h.numericType() == OpenMaya.MFnNumericData.kInt:
						self.appendToResult(h.asInt())
					elif h.numericType() == OpenMaya.MFnNumericData.kFloat:
						self.appendToResult(h.asFloat())
					elif h.numericType() == OpenMaya.MFnNumericData.kDouble:
						self.appendToResult(h.asDouble())
					else:
						print "This sample command only supports boolean, integer, and floating point values."
			advance(adh)

		plug.destructHandle(dh)

# Creator
def cmdCreator():
	return OpenMayaMPx.asMPxPtr(multiPlugInfo())


# Syntax creator
def syntaxCreator():
	syntax = OpenMaya.MSyntax()
	syntax.addFlag(kIndexFlag, kIndexFlagLong, OpenMaya.MSyntax.kNoArg)
	syntax.setObjectType(OpenMaya.MSyntax.kSelectionList, 1, 1)
	return syntax


def initializePlugin(mobject):
	mplugin = OpenMayaMPx.MFnPlugin(mobject, "Autodesk", "1.0", "Any")
	try:
		mplugin.registerCommand(kPluginCmdName, cmdCreator, syntaxCreator)
	except:
		sys.stderr.write( "Failed to register command: %s\n" % kPluginCmdName)
		raise

def uninitializePlugin(mobject):
	mplugin = OpenMayaMPx.MFnPlugin(mobject)
	try:
		mplugin.deregisterCommand(kPluginCmdName)
	except:
		sys.stderr.write("Failed to unregister command: %s\n" % kPluginCmdName)
		raise

File: 2016.sp1/linux/devkit/plug-ins/multiPlugInfoCmd/multiPlugInfoCmd.cpp
//-
// ==========================================================================
// Copyright 2010 Autodesk, Inc. All rights reserved.
//
// Use of this software is subject to the terms of the Autodesk
// license agreement provided at the time of installation or download,
// or which otherwise accompanies this software in either electronic
// or hard copy form.
// ==========================================================================
//+

////////////////////////////////////////////////////////////////////////
//
// multiPlugInfo.cpp
//
// This plugin prints out the child plug information for a multiPlug.
// If the -index flag is used, the logical index values used by the plug
// will be returned.  Otherwise, the plug values will be returned.
//
////////////////////////////////////////////////////////////////////////

#include <maya/MFnPlugin.h>
#include <maya/MPxCommand.h>
#include <maya/MArgDatabase.h>
#include <maya/MArgList.h>
#include <maya/MPlug.h>
#include <maya/MSelectionList.h>
#include <maya/MSyntax.h>

#define kIndexFlag		"-i"
#define kIndexFlagLong	"-index"

///////////////////////////////////////////////////
//
// Command class declaration
//
///////////////////////////////////////////////////
class multiPlugInfo : public MPxCommand
{
public:
					multiPlugInfo();
	virtual			~multiPlugInfo(); 

	MStatus			doIt( const MArgList& args );

	virtual bool	hasSyntax();
	static MSyntax	cmdSyntax();
	static void*	creator();

private:
	MStatus			parseArgs( const MArgList& args );
	MPlug			fPlug;
	bool			isIndex;
};



///////////////////////////////////////////////////
//
// Command class implementation
//
///////////////////////////////////////////////////

multiPlugInfo::multiPlugInfo()
	: isIndex (false)
{}

multiPlugInfo::~multiPlugInfo() {}

void* multiPlugInfo::creator()
{
	return new multiPlugInfo;
}

MSyntax multiPlugInfo::cmdSyntax()
{
	MSyntax syntax;
	syntax.addFlag(kIndexFlag, kIndexFlagLong, MSyntax::kNoArg);
	syntax.setObjectType(MSyntax::kSelectionList, 1, 1);
	syntax.enableQuery( false );
	syntax.enableEdit( false );
	return syntax;
}

bool multiPlugInfo::hasSyntax()
{
	return true;
}

MStatus multiPlugInfo::parseArgs( const MArgList& args )
{
	MStatus status = MS::kSuccess;
	MArgDatabase	argData(syntax(), args, &status);
	if (status != MS::kSuccess)
		return status;

	if (argData.isFlagSet(kIndexFlag))
		isIndex = true;

	// Get the plug specified on the command line.
	MSelectionList slist;
	argData.getObjects(slist);
	if ((slist.length() == 0) ||
		(slist.getPlug(0, fPlug) != MS::kSuccess))
	{
		displayError("Must specify an array plug in the form <nodeName>.<multiPlugName>.", false);
		return MS::kFailure;
	}

	return MS::kSuccess;
}

MStatus multiPlugInfo::doIt( const MArgList& args )
//
// Description
//     This method performs the action of the command.
//
//     This method gets the data stored in the multi attribute
//		and prints it out.
//
{
	if (parseArgs(args) != MS::kSuccess)
		return MS::kFailure;

	// Construct a data handle containing the data stored in the plug.
	MDataHandle dh;
	MStatus stat = fPlug.getValue(dh);
	if (stat != MS::kSuccess)
	{
		displayError("Could not get the plug value.", false);
		return MS::kFailure;
	}
	MArrayDataHandle adh(dh, &stat);
	if (stat != MS::kSuccess)
	{
		displayError("Could not create the array data handle.", false);
		fPlug.destructHandle(dh);
		return MS::kFailure;
	}

	// Iterate over the values in the multiPlug.  If the index flag has been used, just return
	// the logical indices of the child plugs.  Otherwise, return the plug values.
	unsigned int i;
	for (i=0; i<adh.elementCount(); i++, adh.next())
	{
		unsigned int indx = adh.elementIndex(&stat);
		if (stat != MS::kSuccess)
			continue;

		if (isIndex)
			appendToResult((int)indx);
		else
		{
			MDataHandle h = adh.outputValue();
			if (h.isNumeric())
			{
				switch(h.numericType())
				{
					case MFnNumericData::kBoolean:
						appendToResult(h.asBool());
						break;
					case MFnNumericData::kShort:
						appendToResult((int)(h.asShort()));
						break;
					case MFnNumericData::kInt:
						appendToResult(h.asInt());
						break;
					case MFnNumericData::kFloat:
						appendToResult((double)(h.asFloat()));
						break;
					case MFnNumericData::kDouble:
						appendToResult(h.asDouble());
						break;
					default:
						displayError("This sample command only supports boolean, integer, and floating point values.", false);
						stat = MS::kFailure;
						break;
				}
			}
 		}
	}

	fPlug.destructHandle(dh);
	return stat;

}

/////////////////////////////////////////////////////////////////
//
// The following routines are used to register/unregister
// the command we are creating within Maya
//
/////////////////////////////////////////////////////////////////
MStatus initializePlugin( MObject obj )
{
	MStatus   status;
	MFnPlugin plugin( obj, PLUGIN_COMPANY, "1.0", "Any");

	status = plugin.registerCommand( "multiPlugInfo",
									  multiPlugInfo::creator,
									  multiPlugInfo::cmdSyntax);
	if (!status) {
		status.perror("registerCommand");
		return status;
	}

	return status;
}

MStatus uninitializePlugin( MObject obj)
{
	MStatus	  status;
	MFnPlugin plugin( obj );

	status = plugin.deregisterCommand( "multiPlugInfo" );
	if (!status) {
		status.perror("deregisterCommand");
		return status;
	}

	return status;
}
